core/src/main/java/org/infinispan/interceptors/impl/CacheLoaderInterceptor.java

```diff
@@ -221,46 +221,17 @@ public void processEntry(MarshalledEntry marshalledEntry, AdvancedCacheLoader.Ta
    @Override
    public CompletableFuture<Void> visitEntrySetCommand(InvocationContext ctx, EntrySetCommand command)
          throws Throwable {
-      CacheSet<CacheEntry<K, V>> entrySet = (CacheSet<CacheEntry<K, V>>) ctx.forkInvocationSync(command);
-      if (!enabled || hasSkipLoadFlag(command)) {
-         return ctx.shortCircuit(entrySet);
-      }
-      CacheSet<CacheEntry<K, V>> wrappedEntrySet =
-            new AbstractDelegatingEntryCacheSet<K, V>(getCacheWithFlags(cache, command), entrySet) {
-               @Override
-               public CloseableIterator<CacheEntry<K, V>> iterator() {
-                  CloseableIterator<CacheEntry<K, V>> iterator = Closeables.iterator(entrySet.stream());
-            Set<K> seenKeys = new EquivalentHashSet<K>(cache.getAdvancedCache().getDataContainer().size(),
-                              keyEquivalence);
-                  // TODO: how to handle concurrent activation....
-            return new DistinctKeyDoubleEntryCloseableIterator<>(iterator, new CloseableSuppliedIterator<>(
-                              // TODO: how to pass in key filter...
-                    new PersistenceManagerCloseableSupplier<>(executorService, persistenceManager, iceFactory,
-                            new CollectionKeyFilter<>(seenKeys), 10, TimeUnit.SECONDS, 2048)), e -> e.getKey(),
-                    seenKeys);
-               }
-
-               @Override
-               public CloseableSpliterator<CacheEntry<K, V>> spliterator() {
-                  return spliteratorFromIterator(iterator());
-               }
-
-               private <E> CloseableSpliterator<E> spliteratorFromIterator(CloseableIterator<E> iterator) {
-            return new IteratorAsSpliterator.Builder<>(iterator)
-                    .setCharacteristics(Spliterator.CONCURRENT | Spliterator.DISTINCT | Spliterator.NONNULL)
-                    .get();
-               }
-
-               @Override
-               public int size() {
-                  long size = stream().count();
-                  if (size > Integer.MAX_VALUE) {
-                     return Integer.MAX_VALUE;
-                  }
-                  return (int) size;
-               }
-            };
-      return ctx.shortCircuit(wrappedEntrySet);
+      return ctx.onReturn((rCtx, rCommand, rv, throwable) -> {
+         if (throwable != null)
+            throw throwable;
+
+         CacheSet<CacheEntry<K, V>> entrySet = (CacheSet<CacheEntry<K, V>>) rv;
+         if (!enabled || hasSkipLoadFlag(command)) {
+            return CompletableFuture.completedFuture(entrySet);
```

Comment:
actually, you can put everything in the first ```if```, right? if throwable isn't null or not enabled or skip load flags